src/vm/arm/stubs.cpp

```diff
@@ -3782,16 +3782,122 @@ PCODE DynamicHelpers::CreateDictionaryLookupHelper(LoaderAllocator * pAllocator,
     pArgs->module = (CORINFO_MODULE_HANDLE)pModule;
 
     // It's available only via the run-time helper function,
-    // since optimization cases are not yet implemented.
-    assert(pLookup->indirections == CORINFO_USEHELPER);
 
-    BEGIN_DYNAMIC_HELPER_EMIT(18);
+    if (pLookup->indirections == CORINFO_USEHELPER)
+    {
+        BEGIN_DYNAMIC_HELPER_EMIT(18);
 
-    EmitHelperWithArg(p, pAllocator, (TADDR)pArgs, helperAddress);
+        EmitHelperWithArg(p, pAllocator, (TADDR)pArgs, helperAddress);
 
-    END_DYNAMIC_HELPER_EMIT();
+        END_DYNAMIC_HELPER_EMIT();
+    }
+    else
+    {
+        int indirectionsSize = 0;
+        for (WORD i = 0; i < pLookup->indirections; i++)
+        {
+            if ((i == 0 && pLookup->indirectFirstOffset) || (i == 1 && pLookup->indirectSecondOffset))
+            {
+                indirectionsSize += (pLookup->offsets[i] >= 0xFFF ? 10 : 4);
+                indirectionsSize += 4;
+            }
+            else
+            {
+                indirectionsSize += (pLookup->offsets[i] >= 0xFFF ? 10 : 4);
+            }
+        }
+
+        int codeSize = indirectionsSize + (pLookup->testForNull ? 26 : 2);
+
+        BEGIN_DYNAMIC_HELPER_EMIT(codeSize);
+
+        if (pLookup->testForNull)
+        {
+            // mov r3, r0
+            *(WORD *)p = 0x4603;
+            p += 2;
+        }
 
-    // @TODO : Additional implementation is required for optimization cases.
+        for (WORD i = 0; i < pLookup->indirections; i++)
+        {
+            if ((i == 0 && pLookup->indirectFirstOffset) || (i == 1 && pLookup->indirectSecondOffset))
+            {
+                if (pLookup->offsets[i] >= 0xFFF)
+                {
+                    // mov r2, offset
+                    MovRegImm(p, 2, pLookup->offsets[i]);
+                    p += 8;
+
+                    // add r0, r2
+                    *(WORD *)p = 0x4410;
+                    p += 2;
+                }
+                else
+                {
+                    // add r0, r0, <offset>
+                   *(WORD *)(p + 0) = 0xF100;
+                   *(WORD *)(p + 2) = pLookup->offsets[i];
```

Comment:
I'm afraid we have to encode immediate here. According to ARM manual, thumb instruction for `add r0, r0, #imm12` has different encoding. Please check the encoding.

Reply:
Also please use bitmasks and type casting as in below code for `// ldr r0, [r0 + offset]`

Reply:
In addition, instead of using 4 bytes `add r0, r0, #imm12`, we can use 2 bytes `add r0, #imm8`.
What do you think of it?

Reply:
@hqueue OK. I've just copied and pasted, So it made some mistake. Thank you.